cmd/bosun/expr/funcs.go

```diff
@@ -574,39 +618,122 @@ func Band(e *State, T miniprofiler.Timer, query, duration, period string, num fl
 				if e.squelched(res.Tags) {
 					continue
 				}
-				newarr := true
-				for _, a := range r.Results {
-					if !a.Group.Equal(res.Tags) {
-						continue
-					}
-					newarr = false
-					values := a.Value.(Series)
-					for k, v := range res.DPS {
-						i, e := strconv.ParseInt(k, 10, 64)
-						if e != nil {
-							err = e
-							return
-						}
-						values[time.Unix(i, 0).UTC()] = float64(v)
-					}
+				if err = rfunc(r, res); err != nil {
+					return
 				}
-				if newarr {
-					values := make(Series)
-					a := &Result{Group: res.Tags}
-					for k, v := range res.DPS {
-						i, e := strconv.ParseInt(k, 10, 64)
-						if e != nil {
-							err = e
-							return
-						}
-						values[time.Unix(i, 0).UTC()] = float64(v)
-					}
-					a.Value = values
-					r.Results = append(r.Results, a)
+			}
+		}
+	})
+	return
+}
+
+func Window(e *State, T miniprofiler.Timer, query, duration, period string, num float64, rfunc string) (r *Results, err error) {
+	fn, ok := e.GetFunction(rfunc)
+	if !ok {
+		return nil, fmt.Errorf("expr: Window: no %v function", rfunc)
+	}
+	f := reflect.ValueOf(fn.F)
+	r, err = bandTSDB(e, T, query, duration, period, num, func(r *Results, res *opentsdb.Response) error {
+		values := make(Series)
+		min := int64(0)
+		a := &Result{Group: res.Tags}
+		for k, v := range res.DPS {
+			i, e := strconv.ParseInt(k, 10, 64)
+			if e != nil {
+				return e
+			}
+			if min == 0 || i < min {
+				min = i
+			}
+			values[time.Unix(i, 0).UTC()] = float64(v)
+		}
+		if len(values) == 0 {
+			return nil
+		}
+		a.Value = values
```

Comment:
The number of one letter variable names gets real confusing around this point. Perhaps it would be worth investing in some more descriptive names.